### ISSUE TITLE
Identification of 'mail engines'.

### ISSUE THREAD

**Issue by biolithic** [backdrop-contrib/mailsystem#1]
Not too sure how to write the Backdrop equivalent of: 
(because there is no registry table in Backdrop)

```
/**
 * Returns a list of classes which implement MailSystemInterface.
 */
function &mailsystem_get_classes() {
  $mailsystem_classes = &backdrop_static(__FUNCTION__);
  if (!isset($mailsystem_classes)) {
    $mailsystem_classes = array();
    // @todo Is there a better way to find all mail-related classes?
    $declared_classes = get_declared_classes();
    $all_classes = array_combine(
      $declared_classes,
      array_fill(0, count($declared_classes), 0)
    );
    $mail_classes = db_select('registry', 'registry')
      ->distinct()
      ->fields('registry', array('name', 'filename'))
      ->where("type=:type AND ( filename like :filename OR name like :name )",
        // Making the HUGE assumption that all classes which implement
        // MailSystemInterface have filenames containing '.mail.' or
        // classnames ending in 'MailSystem'.
        array(
          ':type' => 'class',
          ':name' => '%MailSystem',
          ':filename' => '%.mail.%',
        )
      )
      ->execute()
      ->fetchAllKeyed();
    foreach ($mail_classes as $classname => $classfile) {
      if ( file_exists($classfile)
        && backdrop_autoload_class($classname)
      ) {
        $all_classes[$classname] = 1;
      }
    }
    foreach ($all_classes as $classname => $autoload) {
      if ( ($autoload || preg_match('/MailSystem/', $classname))
        && ($object = new $classname)
        && ($object instanceof MailSystemInterface)
      ) {
        $mailsystem_classes[$classname] = $classname;
      }
      elseif ($autoload) {
        // Clear classes that are no longer available.
        db_delete('registry')
          ->condition('name', $classname)
          ->execute();
      }
    }
    foreach (array_unique(mailsystem_get()) as $classname) {
      if (class_exists($classname)) {
        $mailsystem_classes[$classname] = $classname;
      }
      else {
        mailsystem_clear(array(mailsystem_default_id() => $classname));
      }
    }
    ksort($mailsystem_classes);
  }
  return $mailsystem_classes;
}
```

any help is appreciated.  Thanks!

**Comment by docwilmot** [comment]
The classes are found via `hook_autoload_info()` now, so you can invoke that hook to find classes ending in `.mail.inc` or having `MailSystem` string in the class name. Note this is the same 'HUGE assumption' made in the original. I still dont know if there is another way.

```
  $mail_classes = array();
  foreach (module_implements('autoload_info') as $module) {
    $function = $module . '_autoload_info';
    $implements = $function();
    foreach($implements as $classname => $path) {
      if(strpos($classname, 'MailSystem') || strrpos($path, 'mail.inc')) {
        $mail_classes[$classname] = $path;
      }
    }
  }
  dpm($mail_classes); 
```

**Comment by biolithic** [comment]
Thanks for the help doc!  This is possibly a shaky solution ("Huge Assumption") but all the mail or newsletter modules depend on this module so I might as well use your solution to get it out the door.  

I have other things to do now but hopefully will try your solution this weekend.

**Comment by Graham-72** [comment]
@biolithic  I am now working on this to try and get it and Mimemail working.

**Comment by Graham-72** [comment]
I have made some revisions to my PR #2 in order to use the config settings for mail systems currently in the site but this is work in progress and by no means a complete solution.

The new method of finding classes is not (for example) finding the mimemail example module, whereas the Drupal version does, so perhaps more thinking is needed. Also I am not sure whether it is possible to get module names rather than class names for the admin form, but perhaps this does not matter.

**Comment by Graham-72** [comment]
I have successfully made use of this module in a real situation with Mime Mail and Views Send. I am not sure that we have the best answer to this particular issue but I am closing it now and maybe a new issue will be appropriate in the future.

**Comment by Graham-72** [comment]
I am re-opening this issue in order to make a new proposal involving site-wide use of hook_mailengine if its parent module Mimemail is installed.

**Comment by Graham-72** [comment]
This is a copy, for the record, of a comment I have posted on gitter.

@herb and any others interested in sending email from Backdrop contributed modules.
I have had an initial look at your port of phpmailer and I am pleased to see that it integrates well with mimemail thanks to its inclusion of hook_mailengine, which is not a hook in core but is provided by the mimemail module so that it and potentially other modules can identify the options for sending mail from the site. I see from Drupal code archives that the SMTP module once included this hook, i.e. provided function smtp_mailengine, but this seems to have been removed at some stage in D7 and is not provided in the current Backdrop release, hence Mimemail cannot identify it as a possible sender of email. I think we should put this right.

In fact I think we should establish clear conventions for Contributed Modules to make use of the power and flexibility provided by the mail class system in core using config system.mail and described in core/includes/mail.inc at line 214 and following. No changes in core are required but we do need small changes in relevant contributed modules, for which I am maintainer for several.

One of the difficulties faced by implementing the Mailsystem and Simple Mailsystem is the identification of which modules in a Backdrop site have the capability of sending mail. This becomes relevant when adding alternatives such as SMTP or PHPMailer. This is currently attempted by identifying any modules that implement a class with ‘MailSystem’ in its name, but this fails for PHPMailer because currently the class name is ‘BackdropPHPMailer’. Although this name could be changed I think a better way forward would be to have a convention that all contributed modules should implement hook_mailengine and, furthermore, this should additionally include a value for a new key ‘mailclass’. This change has no impact on core, a small impact on PHPMailer and a larger one on SMTP. Both Mailsystem and Simple Mailsystem would then make use of this hook to identify mail engines, instead of their current method.

The other convention that should be practised by all modules needing to send email is that they should add at least one entry into the system.mail configuration file. This indicates that they need to send mail and allows the use of Mailsystem or Simple Mailsystem to administer the configuration.

If we adopt this policy it will be possible to configure one module to use the default mail system, another to use PHPMailer, and a third to use Mimemail which in turn is set through its admin menu to use whichever mail engine is thought best. And we have the flexibility to add further mail engines as you have identified.

I am copying this lengthy comment here for future reference and cross-referencing from other contributed modules.
